### PR TITLE
Tracks Opt-Out For Logged Out Users

### PR DESCRIPTION
**Pull request by leomattic**
We are improving the Tracks opt-out mechanism by making sure we do not record any Tracks events if the Analytics option has been disabled in the cookie banner.

Note the opt-out mechanism is already working for logged-in users.

Related to #79719 where we count the number of users opting out of analytics.

## Proposed Changes

* Within the function to record an event
    * Check for the `cookies sensitive_pixel_options`
    * If the Analytics option is turned off, we void recording the event

Something important here is that I am copy-pasting code from `client/lib/analytics/utils/` to `packages/calypso-analytics/src/utils/`. I would love to find a way to avoid the code duplication. Can we import files from there directly?
* `get-tracking-prefs.ts`
* `is-country-in-gdpr-zone.ts`
* `is-region-in-ccpa-zone.ts`

## Testing Instructions

* Run locally
* Make sure you see the cookie banner
* Unselect the Analytics option
* Confirm there are no more Tracks events with Tracks vigilante or watching network traffic

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: config/development.json
Comment (gmovr) [inline config/development.json]: You can also use `?flags=cookie-banner` for testing. Or change these, but they shouldn't be committed to prod. :)

File: packages/calypso-analytics/src/utils/get-tracking-prefs.ts
Comment (gmovr) [inline packages/calypso-analytics/src/utils/get-tracking-prefs.ts]: We have these types here `TrackingPrefs` client/lib/analytics/utils/get-tracking-prefs.ts, is this to move them to a package? If we do, could we also clean up the ones in client?

**Review by gmovr — COMMENTED** [review]
Haven't had time to fully review this, just leaving some notes. 👍

File: config/development.json
Comment (leomattic) [reply config/development.json]: Thanks @gmovr ! I have changed back the option.
`?flags=cookie-banner` is a great tip! 
The setup has been a little confusing for me as I am not currently in a GDPR country and had to try different things to get the banner to show up 😅

File: packages/calypso-analytics/src/utils/get-tracking-prefs.ts
Comment (leomattic) [reply packages/calypso-analytics/src/utils/get-tracking-prefs.ts]: Hi hi @gmovr thanks so much for the review! I am very excited to help this move forward and improve our tracking a little :D 

One very big warning I should make is that I am new to javascript, calypso and to front-end development in general. This is my first PR 🥳 .

I am still getting the hang of components and packages. I think I am missing something obvious here. How can I use TrackingPrefs from here when it's in `client/lib/analytics/utils/` ? I agree we should keep the code in only one place.

Big thanks for the help!!

Comment (robertsreberski) [reply packages/calypso-analytics/src/utils/get-tracking-prefs.ts]: Hey @leomattic! I would be happy to help with that one. You indeed cannot import TrackingPrefs from `client/...`, but we could do it the other way around - it's just so as to not repeat the code.

You can replace uses of TrackingPrefs in `client/lib/analytics/utils/get-tracking-prefs.ts` and other files with `import { TrackingPrefs } from '@automattic/calypso-analytics'`.

File: packages/calypso-analytics/src/tracks.ts
Comment (robertsreberski) [inline packages/calypso-analytics/src/tracks.ts]: As mentioned, we can avoid that by importing the functions we moved to the `calypso-analytics` package in this PR in all use cases in `client/` code

**Review by robertsreberski — CHANGES_REQUESTED** [review]
The code looks good! It's great to see that we improve Tracks analytics to honour user tracking preferences.

The only thing left to do is to clean up the code duplication of methods:
- getTrackingPrefs
- parseTrackingPrefs
- isCountryInGdprZone
- isRegionInCcpaZone

and type `TrackingPrefs`.

As we copied them from `client/lib/analytics` we can remove them from there - and replace all occurrences with ones created in `calypso-analytics` (`import { TrackingPrefs } from '@automattic/calypso-analytics'`).

File: config/development.json
Comment (robertsreberski) [reply config/development.json]: > I am not currently in a GDPR country and had to try different things to get the banner to show up

The whole cookie banner depends on the `country_code` cookie - so you can easily switch countries by modifying it. You can use dev tools to change it to some European country e.g. `NL`.

**Comment by leomattic** [comment]
Hi @robertsreberski 👋  

I removed the files from `client/lib/analytics` and updated the references to `calypso-analytics` as suggested. I would love to get your help with a review as I am not totally sure how things work in javascript 🙏

**Comment by robertsreberski** [comment]
@leomattic the code looks good! thanks for making changes. one last thing to do is to prepare the package for publishing:

https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md#publishing

it basically means to bump the version in package.json of the library code and to understand what actions to take after merging the PR.

File: packages/calypso-analytics/src/tracks.ts
Comment (MicBosi) [reply packages/calypso-analytics/src/tracks.ts]: @leomattic I think we can just remove or update this comment right? Looks like now we don't have code duplication anymore?

File: packages/calypso-analytics/src/utils/get-tracking-prefs.ts
Comment (MicBosi) [inline packages/calypso-analytics/src/utils/get-tracking-prefs.ts]: @leomattic Minor typo: "allo all" -> "allow all"?

File: packages/calypso-analytics/src/tracks.ts
Comment (leomattic) [reply packages/calypso-analytics/src/tracks.ts]: Good catch @MicBosi !!
I am updating it now.

File: packages/calypso-analytics/src/utils/get-tracking-prefs.ts
Comment (leomattic) [reply packages/calypso-analytics/src/utils/get-tracking-prefs.ts]: Very true @MicBosi, thanks!

**Review by MicBosi — APPROVED** [review]
I tested various scenarios, everything works as expected 🎉 

- The user preference gets picked up immediately as soon as the preferences cookie is changed or deleted and Tracks events are fired only when explicit consent is given while in GDPR countries
- Deleting the cookie immediately stops Tracks analytics in GDPR countries
- Changing country code cookie immediately resumes Tracks tracking regardless of preferences cookie settings

Awesome work @leomattic!!!

Many thanks @gmovr and @robertsreberski for helping us out!

@leomattic I guess next step for us is to carefully review and follow the steps in 

https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md#publishing

that Robert mentioned.

**Comment by MicBosi** [comment]
I have updated the package version and changelog. I'm holding off merging atm in order to gather some data from the cookie banner view/ok events and share the results, as well as to make sure the deployment is coordinated with any test or experiment that might be going on.

I've also setup my npmjs account with the A8C organization in order to release the package right after merging.

**Comment by travisw** [comment]
Tried to merge in trunk and it didn't work so well. Will fix.
Meanwhile, please ignore the 5,000+ file changes - that's not intentional.

**Comment by travisw** [comment]
@MicBosi I got this back into a non-conflicted state (after battling git for several hours ⚔️ 😅)

**Comment by travisw** [comment]
> Make sure you see the cookie banner

I'm trying to test this but for some reason am not seeing the cookie banner.
I'm testing as a logged-out user, in a fresh private browser window (no existing cookies), and I'm proxying through from Germany.
I then visit http://calypso.localhost:3000/start (redirects to http://calypso.localhost:3000/start/user-social).

There is no banner that appears.

Should I be seeing it or am I missing something?

**Comment by MicBosi** [comment]
Hi @travisw these are the settings I use. These are for local development, I think there should be a way to launch the branch in the browser and test these settings via url params as well but for me local development seems to work better.

Update `config/develpment.json` to have:

"ad-tracking": true
"cookie-banner": true
then restart using `yarn start`

I use [EditThisCookie](https://chromewebstore.google.com/detail/editthiscookie/fngmhnnpilhplaeedifhccceomclgfbg) to easily inspect and edit cookies.

If you use a VPN you should see the `country_code` reflect that country otherwise you can manually set to something like `GB` or `DE` and the cookie banner should become visible.

Note the banner only becomes visible if the `sensitive_pixel_options` cookie is not set (since it's set by the cookie banner), so if that cookie already exists you need to delete it in order for the cookie banner to appear again.

--- 

I'll continue testing more but at first glance I noticed the Tracks events don't fire while the cookie banner is being displayed, while I think the correct behaviour is for it to fire even if the user hasn't expressed a preference yet, and stop Tracks only if the user explicitly opts-out. 

~~I'm not sure if this was working correctly before the rebase or not though.~~ [edit: see comment below]

**Comment by MicBosi** [comment]
> I'm not sure if this was working correctly before the rebase or not though.

Actually, looks it was my fault to direct Leo to implement it so that Tracks events are blocked unless explicit consent is given, but I believe the correct behaviour is the one followed by the landing pages, ie, fire Tracks unless explicitly disabled.

I'll double check with legal just to have a definitive answer and to make sure this approach is still acceptable.

**Comment by MicBosi** [comment]
@catscheer confirmed the LP behaviour is correct.

**Comment by MicBosi** [comment]
@travisw I've committed the small change necessary to allow Analytics tracking to fire when in GDPR countries when the user hasn't expressed their preference yet - ie, you should be able to see Tracks events being fired while the cookie banner is visible, before clicking on it.

**Comment by travisw** [comment]
Everyting seems to be working as expected and the code makes sense. I also searched around to see if there was anywhere new using any of the exports deleted from `client/lib/analytics/utils/index.ts` and didn't find any.

To confirm that it's the intended functionality, here's what I'm seeing (tested from /plugins):
* Before the cookie banner is accepted, the Tracks events fire.

  Screenshot 2023-12-13 at 11 24 39 AM

* Once I disable the analytics option and accept the banner, then no more Tracks events fire.

  Screenshot 2023-12-13 at 11 24 52 AM

  Screenshot 2023-12-13 at 11 25 52 AM

* If I instead enable the analytics option, Tracks events do fire.

  Screenshot 2023-12-13 at 12 00 35 PM

**Review by travisw — APPROVED** [review]

**Comment by leomattic** [comment]
Woot woot! 🍾
What an adventure!!

So happy to see this across the finish line.
Awesome job team!! 🥇

**Comment by MicBosi** [comment]
1,000,000 thanks to you @leomattic and @travisw 👏 👏 👏

**Pull request merged**
